File: server/src/modules/operation-management/operationController.js
import { queryDb, getDbStatus, inMemoryDb } from '../../config/db.js';

// Update Application Status (Printed, Dispatched, Pending, etc.)
export const updateStatus = async (req, res) => {
  try {
    const { id } = req.params;
    const cleanId = String(id).replace(/^NEX-2026-/, '');
    const { status, remarks } = req.body;

    if (getDbStatus()) {
      await queryDb(
        'UPDATE applications SET status = COALESCE(?, status), remarks = COALESCE(?, remarks) WHERE application_id = ?',
        [status, remarks, cleanId]
      );

      await queryDb(
        'INSERT INTO audit_logs (user_id, action, details) VALUES (?, ?, ?)',
        [req.user ? req.user.user_id : null, 'STATUS_UPDATE', `Application #${cleanId} status updated to ${status || 'unchanged'}`]
      );
    } else {
      const app = (inMemoryDb.applications || []).find(a => String(a.application_id) === String(cleanId) || a.tracking_id === id);
      if (app) {
        if (status) app.status = status;
        if (remarks) app.remarks = remarks;
      }
    }

    return res.status(200).json({ success: true, message: `Application #${cleanId} status updated to ${status}.` });
  } catch (error) {
    return res.status(500).json({ success: false, message: error.message });
  }
};

// Get Print Queue (Applications with status Approved or Processing)
export const getPrintQueue = async (req, res) => {
  try {
    if (getDbStatus()) {
      const rows = await queryDb(`
        SELECT 
          app.application_id,
          CONCAT('NEX-2026-', app.application_id) AS tracking_id,
          app.status,
          app.application_type,
          a.first_name,
          a.last_name,
          CONCAT(a.first_name, ' ', a.last_name) AS fullNameEn,
          a.national_id_number,
          a.date_of_birth AS dob,
          a.gender,
          a.address,
          ic.card_number,
          ic.issue_date,
          ic.expiry_date
        FROM applications app
        JOIN applicants a ON app.applicant_id = a.applicant_id
        LEFT JOIN identity_cards ic ON app.application_id = ic.application_id
        WHERE app.status IN ('Approved', 'Processing', 'Printed')
        ORDER BY app.updated_at DESC
      `);
      return res.status(200).json({ success: true, count: rows.length, queue: rows });
    } else {
      const queue = (inMemoryDb.applications || [])
        .filter(a => ['Approved', 'Processing', 'Printed'].includes(a.status))
        .map(a => ({
          ...a,
          fullNameEn: a.fullNameEn || `${a.first_name || ''} ${a.last_name || ''}`,
          card_number: a.national_id_number || a.nicNumber || 'Pending Issuance'
        }));
      return res.status(200).json({ success: true, count: queue.length, queue });
    }
  } catch (error) {
    return res.status(500).json({ success: false, message: error.message });
  }
};

// ── Record Dispatch: save to dispatch_records table + update application status ──
export const recordDispatch = async (req, res) => {
  try {
    const { id } = req.params;
    const cleanId = String(id).replace(/^NEX-2026-/, '');
    const trackingId = `NEX-2026-${cleanId}`;

    const {
      applicant_name,
      nic_number,
      dispatch_method,   // 'Courier' | 'Postal'
      delivery_address,
      notes
    } = req.body;

    const staffId   = req.user ? req.user.user_id : null;
    const staffName = req.user ? (req.user.full_name || req.user.username) : 'Operational Staff';

    if (getDbStatus()) {
      // 1. Update application status to Dispatched
      await queryDb(
        "UPDATE applications SET status = 'Dispatched' WHERE application_id = ?",
        [cleanId]
      );

      // 2. Insert dispatch record
      await queryDb(
        `INSERT INTO dispatch_records
          (application_id, tracking_id, applicant_name, nic_number, dispatch_method,
           delivery_address, dispatched_by, dispatched_by_name, notes)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          cleanId,
          trackingId,
          applicant_name || 'Unknown',
          nic_number || null,
          dispatch_method || 'Postal',
          delivery_address || null,
          staffId,
          staffName,
          notes || null
        ]
      );

      // 3. Audit log
      await queryDb(
        'INSERT INTO audit_logs (user_id, action, details) VALUES (?, ?, ?)',
        [staffId, 'DISPATCH', `Application ${trackingId} dispatched via ${dispatch_method || 'Postal'} by ${staffName}`]
      );
    } else {
      // In-memory fallback
      const app = (inMemoryDb.applications || []).find(
        a => String(a.application_id) === String(cleanId)
      );
      if (app) app.status = 'Dispatched';

      const nextId = (inMemoryDb.dispatch_records || []).length + 1;
      (inMemoryDb.dispatch_records = inMemoryDb.dispatch_records || []).push({
        dispatch_id:        nextId,
        application_id:     Number(cleanId),
        tracking_id:        trackingId,
        applicant_name:     applicant_name || 'Unknown',
        nic_number:         nic_number || null,
        dispatch_method:    dispatch_method || 'Postal',
        delivery_address:   delivery_address || null,
        dispatched_by:      staffId,
        dispatched_by_name: staffName,
        dispatched_at:      new Date().toISOString(),
        notes:              notes || null
      });
    }

    return res.status(200).json({
      success: true,
      message: `Application ${trackingId} dispatched successfully via ${dispatch_method || 'Postal'}.`
    });
  } catch (error) {
    console.error('recordDispatch error:', error);
    return res.status(500).json({ success: false, message: error.message });
  }
};

// ── Get all Dispatch Records ───────────────────────────────────────────────────
export const getDispatchRecords = async (req, res) => {
  try {
    if (getDbStatus()) {
      const rows = await queryDb(
        `SELECT dispatch_id, application_id, tracking_id, applicant_name, nic_number,
                dispatch_method, delivery_address, dispatched_by, dispatched_by_name,
                dispatched_at, notes
         FROM dispatch_records
         ORDER BY dispatched_at DESC`
      );
      return res.status(200).json({ success: true, count: rows.length, records: rows });
    } else {
      const records = [...(inMemoryDb.dispatch_records || [])].reverse();
      return res.status(200).json({ success: true, count: records.length, records });
    }
  } catch (error) {
    console.error('getDispatchRecords error:', error);
    return res.status(500).json({ success: false, message: error.message });
  }
};

// Get System Analytics & Operational Metrics
export const getAnalytics = async (req, res) => {
  try {
    if (getDbStatus()) {
      const [totalApps] = await queryDb('SELECT COUNT(*) AS total FROM applications');
      const [pendingApps] = await queryDb("SELECT COUNT(*) AS count FROM applications WHERE status = 'Pending'");
      const [approvedApps] = await queryDb("SELECT COUNT(*) AS count FROM applications WHERE status = 'Approved'");
      const [rejectedApps] = await queryDb("SELECT COUNT(*) AS count FROM applications WHERE status = 'Rejected'");
      const [botVerified] = await queryDb("SELECT COUNT(*) AS count FROM applications WHERE bot_verified = 1");
      const [totalUsers] = await queryDb('SELECT COUNT(*) AS count FROM users');

      return res.status(200).json({
        success: true,
        analytics: {
          totalApplications: totalApps[0]?.total || 0,
          pending: pendingApps[0]?.count || 0,
          approved: approvedApps[0]?.count || 0,
          rejected: rejectedApps[0]?.count || 0,
          botVerified: botVerified[0]?.count || 0,
          totalUsers: totalUsers[0]?.count || 0
        }
      });
    } else {
      const apps = inMemoryDb.applications || [];
      return res.status(200).json({
        success: true,
        analytics: {
          totalApplications: apps.length,
          pending: apps.filter(a => a.status === 'Pending').length,
          approved: apps.filter(a => a.status === 'Approved').length,
          rejected: apps.filter(a => a.status === 'Rejected').length,
          botVerified: apps.filter(a => a.bot_verified).length,
          totalUsers: (inMemoryDb.users || []).length
        }
      });
    }
  } catch (error) {
    return res.status(500).json({ success: false, message: error.message });
  }
};

File: server/src/main/java/lk/gov/nexus/controller/OperationController.java
package lk.gov.nexus.controller;

import lk.gov.nexus.config.BackupService;
import lk.gov.nexus.entity.*;
import lk.gov.nexus.repository.*;
import lk.gov.nexus.util.UserContext;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.*;

@RestController
@RequestMapping("/api/operations")
public class OperationController {

    @Autowired private ApplicationRepository applicationRepository;
    @Autowired private ApplicantRepository applicantRepository;
    @Autowired private IdentityCardRepository identityCardRepository;
    @Autowired private VerificationRepository verificationRepository;
    @Autowired private UserRepository userRepository;
    @Autowired private DispatchRecordRepository dispatchRecordRepository;
    @Autowired private AuditLogRepository auditLogRepository;
    @Autowired private BackupService backupService;

    private Long parseId(String id) {
        if (id == null) return null;
        String clean = id.replaceAll("[^0-9]", "");
        return clean.isEmpty() ? null : Long.parseLong(clean);
    }

    // ── Get Operational Metrics & Analytics ───────────────────────────────────
    @GetMapping("/analytics")
    public ResponseEntity<?> getAnalytics() {
        long totalApps = applicationRepository.count();
        long pending = applicationRepository.countByStatus("Pending");
        long approved = applicationRepository.countByStatus("Approved");
        long rejected = applicationRepository.countByStatus("Rejected");
        long botVerified = verificationRepository.countByPassed(1);
        long totalUsers = userRepository.count();

        Map<String, Object> analytics = new LinkedHashMap<>();
        analytics.put("totalApplications", totalApps);
        analytics.put("pending", pending);
        analytics.put("approved", approved);
        analytics.put("rejected", rejected);
        analytics.put("botVerified", botVerified);
        analytics.put("totalUsers", totalUsers);

        return ResponseEntity.ok(Map.of("success", true, "analytics", analytics));
    }

    // ── Get Print Queue (Approved, Processing, Printed) ───────────────────────
    @GetMapping("/print-queue")
    public ResponseEntity<?> getPrintQueue() {
        List<Application> queue = applicationRepository
                .findByStatusInOrderByUpdatedAtDesc(Arrays.asList("Approved", "Processing", "Printed"));

        List<Map<String, Object>> result = new ArrayList<>();
        for (Application app : queue) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("application_id", app.getApplicationId());
            m.put("tracking_id", "NEX-2026-" + app.getApplicationId());
            m.put("status", app.getStatus());
            m.put("application_type", app.getApplicationType());

            if (app.getApplicantId() != null) {
                applicantRepository.findById(app.getApplicantId()).ifPresent(applicant -> {
                    m.put("first_name", applicant.getFirstName());
                    m.put("last_name", applicant.getLastName());
                    m.put("fullNameEn", applicant.getFirstName() + " " + applicant.getLastName());
                    m.put("national_id_number", applicant.getNationalIdNumber());
                    m.put("dob", applicant.getDateOfBirth() != null ? applicant.getDateOfBirth().toString() : null);
                    m.put("gender", applicant.getGender());
                    m.put("address", applicant.getAddress());
                });
            }

            identityCardRepository.findByApplicationId(app.getApplicationId()).ifPresent(card -> {
                m.put("card_number", card.getCardNumber());
                m.put("issue_date", card.getIssueDate() != null ? card.getIssueDate().toString() : null);
                m.put("expiry_date", card.getExpiryDate() != null ? card.getExpiryDate().toString() : null);
            });

            if (!m.containsKey("card_number")) {
                m.put("card_number", m.get("national_id_number") != null ? m.get("national_id_number") : "Pending Issuance");
            }

            result.add(m);
        }

        return ResponseEntity.ok(Map.of("success", true, "count", result.size(), "queue", result));
    }

    // ── Update Application Status ─────────────────────────────────────────────
    @RequestMapping(value = {"/applications/{id}/status", "/{id}/status"}, method = {RequestMethod.PATCH, RequestMethod.PUT})
    public ResponseEntity<?> updateStatus(@PathVariable String id, @RequestBody Map<String, String> body) {
        Long appId = parseId(id);
        if (appId == null) return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Invalid ID"));

        Optional<Application> appOpt = applicationRepository.findById(appId);
        if (appOpt.isEmpty()) return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("success", false, "message", "Application not found"));

        Application app = appOpt.get();
        String status = body.get("status");
        String remarks = body.get("remarks");

        if (status != null) app.setStatus(status);
        if (remarks != null) app.setRemarks(remarks);
        app.setUpdatedAt(LocalDateTime.now());
        applicationRepository.save(app);

        auditLogRepository.save(new AuditLog(UserContext.getCurrentUserId(), "STATUS_UPDATE",
                "Application #" + appId + " status updated to " + (status != null ? status : "unchanged")));
        backupService.triggerAutoBackup("Application #" + appId + " status updated to " + status);

        return ResponseEntity.ok(Map.of("success", true, "message", "Application #" + appId + " status updated to " + status + "."));
    }

    // ── Record Dispatch ───────────────────────────────────────────────────────
    @PostMapping("/dispatch/{id}")
    public ResponseEntity<?> recordDispatch(@PathVariable String id, @RequestBody Map<String, String> body) {
        Long appId = parseId(id);
        if (appId == null) return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Invalid ID"));

        String trackingId = "NEX-2026-" + appId;
        String applicantName = body.get("applicant_name") != null ? body.get("applicant_name") : "Unknown";
        String nicNumber = body.get("nic_number");
        String dispatchMethod = body.get("dispatch_method") != null ? body.get("dispatch_method") : "Postal";
        String deliveryAddress = body.get("delivery_address");
        String notes = body.get("notes");

        Long staffId = UserContext.getCurrentUserId();
        String staffName = UserContext.getCurrentUser() != null ? UserContext.getCurrentUser().getFullName() : "Operational Staff";

        // 1. Update application status to Dispatched
        applicationRepository.findById(appId).ifPresent(app -> {
            app.setStatus("Dispatched");
            app.setUpdatedAt(LocalDateTime.now());
            applicationRepository.save(app);
        });

        // 2. Insert dispatch record
        DispatchRecord record = new DispatchRecord();
        record.setApplicationId(appId);
        record.setTrackingId(trackingId);
        record.setApplicantName(applicantName);
        record.setNicNumber(nicNumber);
        record.setDispatchMethod(dispatchMethod);
        record.setDeliveryAddress(deliveryAddress);
        record.setDispatchedBy(staffId);
        record.setDispatchedByName(staffName);
        record.setDispatchedAt(LocalDateTime.now());
        record.setNotes(notes);
        dispatchRecordRepository.save(record);

        auditLogRepository.save(new AuditLog(staffId, "DISPATCH",
                "Application " + trackingId + " dispatched via " + dispatchMethod + " by " + staffName));
        backupService.triggerAutoBackup("Application " + trackingId + " dispatched");

        return ResponseEntity.ok(Map.of(
                "success", true,
                "message", "Application " + trackingId + " dispatched successfully via " + dispatchMethod + "."
        ));
    }

    // ── Get All Dispatch Records ──────────────────────────────────────────────
    @GetMapping("/dispatch-records")
    public ResponseEntity<?> getDispatchRecords() {
        List<DispatchRecord> list = dispatchRecordRepository.findAllByOrderByDispatchedAtDesc();
        return ResponseEntity.ok(Map.of("success", true, "count", list.size(), "records", list));
    }
}
